api/src/paths/artifact/intake.ts

```diff
@@ -0,0 +1,188 @@
+              format: 'uuid',
+              description:
+                'The unique identifier for this artifact collection.'
+            }
+          }
+        }
+      }
+    }
+  },
+  responses: {
+    200: {
+      description: 'Submission OK.',
+      content: {
+        'application/json': {
+          schema: {
+            type: 'object',
+            
+          }
+        }
+      }
+    },
+    ...defaultErrorResponses
+  }
+};
+
+export function intakeArtifacts(): RequestHandler {
+  return async (req, res) => {
+    if (!req?.files?.length) {
+      throw new HTTP400('Missing required `media`');
+    }
+
+    const files: Express.Multer.File[] = Object.values(req.files);
+    if (req.body.metadata?.length !== files.length) {
+      throw new HTTP400('File count and metadata record count do not agree');
+    }
+
+    const dataPackageId = req.body?.data_package_id;
+    if (!dataPackageId) {
+      throw new HTTP400('Data package ID is required');
+    }
+
+    // Scan all files for viruses
+    const virusScanResults = await Promise.all(files.map(async (file) => scanFileForVirus(file)));
+    if (virusScanResults.some((result: boolean) => !result)) {
+      throw new HTTP400('Malicious content detected, upload cancelled');
+    }
+
+    const sourceSystem = getKeycloakSource(req['keycloak_token']);
+    if (!sourceSystem) {
+      throw new HTTP400('Failed to identify known submission source system', [
+        'token did not contain a clientId/azp or clientId/azp value is unknown'
+      ]);
+    }
+
+    const connection = getServiceAccountDBConnection(sourceSystem);
+
+    try {
+      await connection.open();
+
+      const submissionService = new SubmissionService(connection);
+      const artifactService = new ArtifactService(connection);
+
+      // Fetch the source transform record for this submission based on the source system user id
+      const sourceTransformRecord = await submissionService.getSourceTransformRecordBySystemUserId(
+        connection.systemUserId()
+      );
+
+      if (!sourceTransformRecord) {
+        throw new HTTP400('Failed to get source transform record for system user');
+      }
+
+      const metadataRecords: string[] = req.body.metadata;
+      const { source_transform_id } = sourceTransformRecord;
+
+      // Create a new submission for the artifact collection
+      const { submission_id } = await submissionService.getOrInsertSubmissionRecord({
+        source_transform_id: sourceTransformRecord.source_transform_id,
+        uuid: dataPackageId
+      });
+      
+      // Retrieve the ID that will be assigned to the artifact once it's inserted
+      const insertArtifactId = await artifactService.getNextArtifactId();
+
+      const artifactPersistPromises = files.map(async (file: Express.Multer.File, fileIndex: number) => {
+        const artifactMetadata: IArtifactMetadata = JSON.parse(metadataRecords[fileIndex]);
+        const artifactId = insertArtifactId + fileIndex;
```

Comment:
Theoretically this should be safe against collisions. Can't do an `await artifactService.getNextArtifactId()` because until the transaction finishes, this resolves the same artifact ID.

Reply:
Actually, this won't work if we use `artifact_seq`, because it wouldn't increment the sequence

Reply:
If we swap the SQL to use the sequence, then we won't be able to do this, as the sequence won't have incremented enough times.

I'm not sure if there are any downsides, but this seems to be a potential solution:
`select nextval('your_sequence') from generate_series(1,<number of values to return>);`

Reply:
That came up online as a way of fetching multiple values from a sequence. I can't speak to any pros/cons about it though.

Reply:
> I'm not sure if there are any downsides, but this seems to be a potential solution:
`select nextval('your_sequence') from generate_series(1,<number of values to return>);`

That seems promising - I'll give it a try